### PR TITLE
ACIModule: Add new methods to dynamically build URLs

### PR DESCRIPTION
**Pull request by jmcgill298**
@dagwieers Travis will probably complain, I will clean it up later.

I have added the `ACIModule.get_url()` method to be used in the modules. The method should pass the identerfiers per what is used in the modules. If there are 1-to-1 mappings, then their classes should be passed as `child_classes`. I have included aci_tenant and aci_bd as example files, but here is how aci_bd would look `aci.get_url(root_class="tenant", subclass_1="bd", child_classes=("fvRsCtx", etc.))`.

The only requirement for adding new classes is to update the `URL_MAPPING` dict in aci.py. This dict should save some work in having to look up URL info for class, mo, and root keys.

Once the object is 3 levels deep, the possible options are pretty large. I will have to test these tomorrow (I test objects up to 2 levels with children). 

I have not developed 4 levels deep, but 'present' and 'absent' should work, and 'query' will currently just get all objects of the class.

`aci_subnet` will be an exception for now as the root key is an IP Address which needs to have `[]` around it.

### CODE REVIEW

File: library/aci_bd.py [hunk 1]
```diff
@@ -219,23 +219,25 @@ def main():
...
+    # elif tenant is not None:
+    #     path = 'api/mo/uni/tn-%(tenant)s.json' % module.params
+    #     filter_string = '?rsp-subtree=full&rsp-subtree-class=fvBD,fvRsCtx,fvRsIgmpsn,fvRsBDToNdP,fvRsBdToEpRet'
+    # else:
+    #     path = 'api/class/fvBD.json'
+    #     filter_string = "?rsp-subtree=full&rsp-subtree-class=fvBD,fvRsCtx,fvRsIgmpsn,fvRsBDToNdP,fvRsBdToEpRet"
 
     aci = ACIModule(module)
 
+    path, filter_string = aci.get_url(root_class="tenant", subclass_1="bd", child_classes=('fvRsCtx', 'fvRsIgmpsn', 'fvRsBDToNdP', 'fvRsBdToEpRet'))
```
Comment (dagwieers) [inline library/aci_bd.py]: So I have two things I don't like:
- The method **get_url** somehow resembles the Ansible module by the same name, and method **fetch_url** (both download a URL), I'd rather use something like **construct_url**.
- I don't like the **subclass_1** scheme, why don't you make it a list ? `subclass=('bd')`

This reflects also the implementation with the various get_url_X methods.

Comment (jmcgill298) [inline library/aci_bd.py]: I didn't like the name either, `construct_url` is better

I like the explicitness of the scheme; it is clearer that it is ordered and how many are supported, but I'm not set on it. I will make it a list.

Comment (dagwieers) [inline library/aci_bd.py]: We already have `define_protocol` and `define_method`, and both of them update either self.params or self. Maybe we could do the same with a `define_url` or `define_path` method. So we don't have to pass that information around.

Comment (jmcgill298) [inline library/aci_bd.py]: I was planning on doing this once all the modules are converted over, which will change how the URLs are retrieved by the methods used to interact with the APIC.

**Comment by jmcgill298** [comment]
@dagwieers Please review again and let me know of any updates that need to happen before merging.

All modules in this PR have been migrated and tested using the new method.

Also, please note that I have renamed the "app_profile" param to "ap" to match the module. I can revert this, but it seemed odd to have the module name "ap" and the object it references named "app_profile" and also seemed inconsistent with "bd", "epg", etc.

**Comment by dagwieers** [comment]
Maybe we should get this merged sooner rather than later. Because one  large PR is going to be harder to accept. I'd rather decoupled the library (and maybe one example module) from all the other modules.

**Comment by jmcgill298** [comment]
added individually

**Pull request closed without merging**